hazelcast-client/src/main/java/com/hazelcast/client/cache/impl/NearCachedClientCacheProxy.java

```diff
@@ -219,123 +218,139 @@ protected void onPutIfAbsentAsyncInternal(K key, V value, Data keyData, Data val
+            dataKeys.add(keyData);
+        }
+        return keyMap;
+    }
+
+    private void populateResultFromNearCache(Collection<?> keys, List<Object> resultingKeyValuePairs) {
         Iterator<?> iterator = keys.iterator();
         while (iterator.hasNext()) {
             Object key = iterator.next();
             Object cached = getCachedValue(key, true);
             if (cached != NOT_CACHED) {
-                result.put((K) (serializeKeys ? toObject(key) : key), (V) cached);
+                resultingKeyValuePairs.add(key);
+                resultingKeyValuePairs.add(cached);
                 iterator.remove();
             }
         }
     }
 
-    private void populateResultFromRemote(Map<K, V> resultMap, Map<Object, Long> reservations,
-                                          List<Map.Entry<Data, Data>> response) {
-        if (serializeKeys) {
-            for (Map.Entry<Data, Data> entry : response) {
-                Data keyData = entry.getKey();
-                Long reservationId = reservations.get(keyData);
-                if (reservationId != null) {
-                    K key = toObject(keyData);
-                    V remoteValue = resultMap.get(key);
-                    V cachedValue = tryPublishReserved(keyData, remoteValue, reservationId);
-                    V newValue = toObject(cachedValue);
-                    resultMap.put(key, newValue);
-                    reservations.remove(keyData);
-                }
+    private Map<Object, Long> getNearCacheReservations(Collection<?> nearCacheKeys, Map<Object, Data> keyMap) {
+        Map<Object, Long> reservations = createHashMap(nearCacheKeys.size());
+        for (Object key : nearCacheKeys) {
+            Data keyData = serializeKeys ? (Data) key : keyMap.get(key);
+            long reservationId = tryReserveForUpdate(key, keyData);
+            if (reservationId != NOT_RESERVED) {
+                reservations.put(key, reservationId);
             }
-        } else {
-            for (Map.Entry<K, V> entry : resultMap.entrySet()) {
-                K key = entry.getKey();
-                V value = entry.getValue();
-                Long reservationId = reservations.get(key);
-                if (reservationId != null) {
-                    V cachedValue = tryPublishReserved(key, value, reservationId);
-                    V newValue = toObject(cachedValue);
-                    resultMap.put(key, newValue);
-                    reservations.remove(key);
-                }
+        }
+        return reservations;
+    }
+
+    private void populateResultFromRemote(int currentSize, List<Object> resultingKeyValuePairs, Map<Object, Long> reservations) {
+        for (int i = currentSize; i < resultingKeyValuePairs.size(); i += 2) {
+            Data keyData = (Data) resultingKeyValuePairs.get(i);
+            Data valueData = (Data) resultingKeyValuePairs.get(i + 1);
+
+            Object ncKey = serializeKeys ? keyData : toObject(keyData);
```

Comment:
I had the same thought here. It's not just the deserialization cost, it's also the issue with the OBJECT key not being the same instance as the original one. So we rely on `equals()`/`hashCode()` here to have `Long reservationId = reservations.get(ncKey);` working properly. This is actually a real issue, since it will fail silently otherwise and never populate the Near Cache -> super hard to debug when a customer reports a "Near Cache is not working". A bi-directional map would solve this bigger issue as well.